File: src/kernel/arch/host/Platform.cc
/*
 * Copyright (C) 2016-2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Config.h>
#include <base/log/Kernel.h>
#include <base/Init.h>

#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <vector>

#include "mem/MainMemory.h"
#include "Args.h"
#include "TCU.h"
#include "Platform.h"

namespace kernel {

m3::BootInfo::Mod *Platform::_mods;
m3::PEDesc *Platform::_pes;
m3::BootInfo::Mem *Platform::_mems;
m3::BootInfo Platform::_info;

static MainMemory::Allocation binfomem;

void Platform::init() {
    size_t cores = PE_COUNT;
    const char *cores_str = getenv("M3_CORES");
    if(cores_str) {
        cores = strtoul(cores_str, NULL, 10);
        if(cores < 2 || cores > PE_COUNT)
            PANIC("Invalid PE count (min=2, max=" << PE_COUNT << ")");
    }

    // init PEs
    size_t total_pes = cores + 1;
    if(Args::bridge)
        total_pes += 2;
    if(Args::disk)
        total_pes++;
    _info.pe_count = total_pes;
    _pes = new m3::PEDesc[total_pes];
    size_t i = 0;
    for(; i < cores; ++i)
        _pes[i] = m3::PEDesc(m3::PEType::COMP_IMEM, m3::PEISA::X86, 1024 * 1024);

    // these are dummy PEs; they do not really exist, but serve the purpose to let root not
    // complain that the IDE/NIC PE isn't present.
    if(Args::bridge) {
        _pes[i++] = m3::PEDesc(m3::PEType::COMP_IMEM, m3::PEISA::NIC, 0);
        _pes[i++] = m3::PEDesc(m3::PEType::COMP_IMEM, m3::PEISA::NIC, 0);
    }
    if(Args::disk)
        _pes[i++] = m3::PEDesc(m3::PEType::COMP_IMEM, m3::PEISA::IDE_DEV, 0);

    _pes[i++] = m3::PEDesc(m3::PEType::MEM, m3::PEISA::NONE, TOTAL_MEM_SIZE);

    // create memory
    uintptr_t base = reinterpret_cast<uintptr_t>(
        mmap(0, TOTAL_MEM_SIZE, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0));

    if(TOTAL_MEM_SIZE <= FS_MAX_SIZE + Args::kmem)
        PANIC("Not enough DRAM");

    MainMemory &mem = MainMemory::get();
    mem.add(new MemoryModule(MemoryModule::OCCUPIED, m3::GlobAddr(0, base), FS_MAX_SIZE));
    mem.add(new MemoryModule(MemoryModule::KERNEL, m3::GlobAddr(0, base + FS_MAX_SIZE), Args::kmem));
    size_t usize = TOTAL_MEM_SIZE - (FS_MAX_SIZE + Args::kmem);
    mem.add(new MemoryModule(MemoryModule::USER, m3::GlobAddr(0, base + FS_MAX_SIZE + Args::kmem), usize));

    // set memories
    _info.mem_count = 2;
    _mems = new m3::BootInfo::Mem[2];
    _mems[0] = m3::BootInfo::Mem(0, FS_MAX_SIZE, true);
    _mems[1] = m3::BootInfo::Mem(FS_MAX_SIZE + Args::kmem, usize, false);
}

void Platform::add_modules(int argc, char **argv) {
    MainMemory &mem = MainMemory::get();

    _mods = new m3::BootInfo::Mod[argc];

    size_t bmodsize = 0;
    for(int i = 0; i < argc; ++i) {
        m3::OStringStream args;
        args << basename(argv[i]);

        strcpy(_mods[i].name, args.str());

        bmodsize += sizeof(m3::BootInfo::Mod) + args.length() + 1;

        // copy boot module into memory
        {
            int fd = open(argv[i], O_RDONLY);
            if(fd < 0)
                PANIC("Opening '" << argv[i] << "' for reading failed");
            struct stat info;
            if(fstat(fd, &info) == -1)
                PANIC("Stat for '" << argv[i] << "' failed");

            MainMemory::Allocation alloc = mem.allocate(static_cast<size_t>(info.st_size), 1);
            if(!alloc)
                PANIC("Not enough memory for boot module '" << argv[i] << "'");
            ssize_t res = read(fd, reinterpret_cast<void*>(alloc.addr().offset()), alloc.size);
            if(res == -1)
                PANIC("Reading from '" << argv[i] << "' failed");
            close(fd);

            _mods[i].addr = alloc.addr().offset();
            _mods[i].size = alloc.size;
        }
    }

    // set modules
    _info.mod_count = static_cast<uint64_t>(argc);

    // build kinfo page
    size_t bsize = sizeof(m3::BootInfo) + _info.mod_count * sizeof(m3::BootInfo::Mod)
                                        + _info.pe_count * sizeof(m3::BootInfo::PE)
                                        + _info.mem_count * sizeof(m3::BootInfo::Mem);
    binfomem = mem.allocate(bsize, 1);
    if(!binfomem)
        PANIC("Not enough memory for boot info");
    m3::BootInfo *binfo = reinterpret_cast<m3::BootInfo*>(binfomem.addr().offset());
    memcpy(binfo, &_info, sizeof(_info));
    binfo->pe_count -= 2;

    // add modules
    uintptr_t mod_addr = binfomem.addr().offset() + sizeof(_info);
    for(uint64_t i = 0; i < _info.mod_count; ++i) {
        memcpy(reinterpret_cast<void*>(mod_addr), _mods + i, sizeof(_mods[i]));
        mod_addr += sizeof(_mods[i]);
    }

    // add PEs
    for(uint64_t i = 1; i < _info.pe_count - 1; ++i) {
        m3::BootInfo::PE pe;
        pe.id = i;
        pe.desc = Platform::_pes[i];
        memcpy(reinterpret_cast<void*>(mod_addr), &pe, sizeof(pe));
        mod_addr += sizeof(pe);
    }

    // add memory regions
    for(uint64_t i = 0; i < _info.mem_count; ++i) {
        memcpy(reinterpret_cast<void*>(mod_addr), _mems + i, sizeof(_mems[i]));
        mod_addr += sizeof(_mems[i]);
    }
}

m3::GlobAddr Platform::info_addr() {
    return binfomem.addr();
}

peid_t Platform::kernel_pe() {
    return 0;
}
peid_t Platform::first_pe() {
    return 1;
}
peid_t Platform::last_pe() {
    return _info.pe_count - 2;
}

bool Platform::is_shared(peid_t) {
    return false;
}

goff_t Platform::rbuf_pemux(peid_t) {
    // unused
    return 0;
}

}

File: src/rustkernel/src/arch/host/platform.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use base::cfg;
use base::col::{String, Vec};
use base::goff;
use base::kif::{boot, PEDesc, PEType, PEISA};
use base::libc;
use base::tcu::PEId;
use base::util;
use core::mem::MaybeUninit;
use core::ptr;

use args;
use mem;
use platform;

pub fn init(args: &[String]) -> platform::KEnv {
    let mut info = boot::Info::default();

    // PEs
    let mut pes = Vec::new();
    for _ in 0..cfg::PE_COUNT {
        pes.push(PEDesc::new(PEType::COMP_IMEM, PEISA::X86, 1024 * 1024));
    }
    if args::get().disk {
        pes.push(PEDesc::new(PEType::COMP_IMEM, PEISA::IDE_DEV, 0));
    }
    if args::get().net_bridge.is_some() {
        pes.push(PEDesc::new(PEType::COMP_IMEM, PEISA::NIC_DEV, 0));
        pes.push(PEDesc::new(PEType::COMP_IMEM, PEISA::NIC_DEV, 0));
    }
    let mut upes = Vec::new();
    for (i, pe) in pes[1..].iter().enumerate() {
        upes.push(boot::PE::new(i as u32, *pe));
    }
    info.pe_count = upes.len() as u64;

    let mems = build_mems();
    info.mem_count = mems.len() as u64;

    let mods = build_modules(args);
    info.mod_count = mods.len() as u64;

    // build kinfo page
    let bsize = util::size_of::<boot::Info>()
        + info.mod_count as usize * util::size_of::<boot::Mod>()
        + info.pe_count as usize * util::size_of::<boot::PE>()
        + info.mem_count as usize * util::size_of::<boot::Mem>();
    let mut binfo_mem = mem::get()
        .allocate(bsize as goff, 1)
        .expect("Unable to allocate mem for boot info");

    unsafe {
        // info
        let mut dest = binfo_mem.global().offset();
        libc::memcpy(
            dest as *mut u8 as *mut libc::c_void,
            &info as *const boot::Info as *const libc::c_void,
            util::size_of::<boot::Info>(),
        );
        dest += util::size_of::<boot::Info>() as goff;

        // modules
        libc::memcpy(
            dest as *mut u8 as *mut libc::c_void,
            mods.as_ptr() as *const libc::c_void,
            mods.len() * util::size_of::<boot::Mod>(),
        );
        dest += (mods.len() * util::size_of::<boot::Mod>()) as goff;

        // PEs
        libc::memcpy(
            dest as *mut u8 as *mut libc::c_void,
            upes.as_ptr() as *const libc::c_void,
            upes.len() * util::size_of::<boot::PE>(),
        );
        dest += (upes.len() * util::size_of::<boot::PE>()) as goff;

        // memories
        libc::memcpy(
            dest as *mut u8 as *mut libc::c_void,
            mems.as_ptr() as *const libc::c_void,
            mems.len() * util::size_of::<boot::Mem>(),
        );
    }
    binfo_mem.claim();

    platform::KEnv::new(info, binfo_mem.global(), mods, pes)
}

fn build_mems() -> Vec<boot::Mem> {
    // create memory
    let base = unsafe {
        libc::mmap(
            ptr::null_mut(),
            cfg::TOTAL_MEM_SIZE,
            libc::PROT_READ | libc::PROT_WRITE,
            libc::MAP_ANON | libc::MAP_PRIVATE,
            -1,
            0,
        )
    };
    assert!(base != libc::MAP_FAILED);
    let mut off = base as goff;

    // fs image
    mem::get().add(mem::MemMod::new(
        mem::MemType::OCCUPIED,
        kernel_pe(),
        off,
        cfg::FS_MAX_SIZE as goff,
    ));
    off += cfg::FS_MAX_SIZE as goff;

    // kernel memory
    mem::get().add(mem::MemMod::new(
        mem::MemType::KERNEL,
        kernel_pe(),
        off,
        args::get().kmem as goff,
    ));
    off += args::get().kmem as goff;

    // user memory
    let user_size = cfg::TOTAL_MEM_SIZE - (cfg::FS_MAX_SIZE + args::get().kmem);
    mem::get().add(mem::MemMod::new(
        mem::MemType::USER,
        kernel_pe(),
        off,
        user_size as goff,
    ));

    // set memories
    let mut mems = Vec::new();
    mems.push(boot::Mem::new(0, cfg::FS_MAX_SIZE as goff, true));
    mems.push(boot::Mem::new(off, user_size as goff, false));
    mems
}

fn build_modules(args: &[String]) -> Vec<boot::Mod> {
    let mut mods = Vec::new();

    for arg in args {
        // copy boot module into memory
        unsafe {
            let fd = libc::open(arg.as_ptr() as *const libc::c_char, libc::O_RDONLY);
            if fd == -1 {
                panic!("Opening {} for reading failed", arg);
            }
            let mut finfo: libc::stat = MaybeUninit::uninit().assume_init();
            if libc::fstat(fd, &mut finfo) == -1 {
                panic!("Stat for {} failed", arg);
            }

            let mut alloc = mem::get()
                .allocate(finfo.st_size as goff, 1)
                .expect("Unable to alloc mem for boot module");
            let dest = alloc.global().offset() as *mut u8 as *mut libc::c_void;
            if libc::read(fd, dest, alloc.size() as usize) == -1 {
                panic!("Reading from {} failed", arg);
            }
            libc::close(fd);

            let mod_name = arg.rsplitn(2, '/').next().unwrap();
            mods.push(boot::Mod::new(alloc.global().raw(), alloc.size(), mod_name));

            // don't free mem
            alloc.claim();
        }
    }

    mods
}

pub fn kernel_pe() -> PEId {
    0
}
pub fn user_pes() -> platform::PEIterator {
    platform::PEIterator::new(1, platform::pes().len() - 1)
}

pub fn is_shared(_pe: PEId) -> bool {
    false
}

pub fn rbuf_pemux(_pe: PEId) -> goff {
    0
}
